File: 02Rasterization/ellipse.py
import matplotlib as mpl
import matplotlib.pyplot as plt
import numpy as np
import math

def eulerMatrix(theta):
    return np.array([[math.cos(theta),-math.sin(theta)],[math.sin(theta),math.cos(theta)]])

a,b,theta=[float(x) for x in input().split()]
slices=100
x=math.cos(theta)*a
y=math.sin(theta)*a
dphi=2*math.pi/slices
r=np.array([x,y]).reshape((2,1))
R=eulerMatrix(theta)
A=np.array([[math.cos(dphi),-a/b*math.sin(dphi)],[b/a*math.sin(dphi),math.cos(dphi)]])
# be careful with the transpose
M=np.dot(R,np.dot(A,R.T))

pointList=[]
plt.scatter(*r)
for phi in np.linspace(0,2*math.pi,slices):
    r=np.dot(M,r)
    pointList.append(np.round(r))
    plt.scatter(*pointList[-1])
plt.show()
# print(pointList)
# print(M)

# Another version:

# r=np.array([a,0]).reshape((2,1))
# plt.scatter(*np.dot(R,r))
# for phi in np.linspace(0,2*math.pi,slices):
#     r=np.dot(A,r)
#     plt.scatter(*np.dot(R,r))
# plt.show()

# draw a filled a polygon.

# def fill(pointCloud):
#     """
#     pass in a N*2 point array as an ordered polygon.
#     """
#     for y in range(np.min(pointCloud(:,1)),np.max(pointCloud(:,1))+1):
#         for x in activeRange()
    

File: 02Rasterization/ellipse.cpp
#include "ellipse.h"
// #define _USE_MATH_DEFINES
#include <cmath>


Polygon ellipse(const double a, const double b,const Point& center,
                                             const double theta) {
    double cosine = std::cos(theta), sine = std::sin(theta);
    double R[4] = {cosine, -sine, sine, cosine};
    const int SLICES = 100;
    const double PI = 3.14159265358979323846264338327950288,dphi = PI / SLICES;
    
    cosine = std::cos(dphi);
    sine = std::sin(dphi);
    double A[4] = {cosine,-a/b*sine,b/a*sine,cosine},M[4];
    // Matrix Multiplication. B=R A (R.T)
    for (int i=0;i<2;i++)
        for (int j=0;j<2;j++){
            M[i*2+j]=0;
            for (int k=0;k<2;k++)
                for (int l=0;l<2;l++)
                    M[i*2+j]+=R[i*2+l]*A[l*2+k]*R[j*2+k];
        }

    Polygon points;
    points.reserve(SLICES);

    double x=a*R[0],y=a*R[2];
    points.push_back(Point(x,y)+center);
    for (double phi=0;phi<2*PI;phi+=dphi){
        x=M[0]*x+M[1]*y;
        y=M[2]*x+M[3]*y;
        points.push_back(Point(x,y)+center);
    }
    return points;
}